src/query/insert.rs

```diff
@@ -155,9 +155,67 @@ where
         M: IntoActiveModel<A>,
         I: IntoIterator<Item = M>,
     {
+        let mut columns: Vec<_> = <A::Entity as EntityTrait>::Column::iter()
+            .map(|_| None)
+            .collect();
+        let mut null_value: Vec<Option<Value>> =
+            std::iter::repeat(None).take(columns.len()).collect();
+        let mut all_values: Vec<Vec<SimpleExpr>> = Vec::new();
+
         for model in models.into_iter() {
-            self = self.add(model);
+            let mut am: A = model.into_active_model();
+            self.primary_key =
+                if !<<A::Entity as EntityTrait>::PrimaryKey as PrimaryKeyTrait>::auto_increment() {
+                    am.get_primary_key_value()
+                } else {
+                    None
+                };
+            let mut values = Vec::with_capacity(columns.len());
+            for (idx, col) in <A::Entity as EntityTrait>::Column::iter().enumerate() {
+                let av = am.take(col);
+                match av {
+                    ActiveValue::Set(value) | ActiveValue::Unchanged(value) => {
+                        columns[idx] = Some(col); // mark the column as used
+                        null_value[idx] = Some(value.as_null()); // store the null value with the correct type
+                        values.push(col.save_as(Expr::val(value))); // same as add() above
+                    }
+                    ActiveValue::NotSet => {
+                        values.push(SimpleExpr::Keyword(Keyword::Null)); // indicate a missing value
+                    }
+                }
+            }
+            all_values.push(values);
         }
+
+        if !all_values.is_empty() {
+            // filter only used column
+            self.query.columns(columns.iter().cloned().flatten());
+
+            // flag used column
+            for col in columns.iter() {
+                self.columns.push(col.is_some());
+            }
```

Comment:
thanks. self.columns is really for use with add(). so should we do `self.columns.clear()` first?

Reply:
Okay, reset (clear) it before altering it seems safer.

Reply:
Or simply replace the columns vector, isn't this more readable?

```rs
self.columns = columns.iter().map(Option::is_some).collect();
```